src/Game/AccountManagement.java

```diff
@@ -114,6 +123,222 @@ public static void register(String user, String pass) {
+			int idx;
+			for (int var12 = 0; var12 < 5; ++var12) {
+				int length = StreamUtil.readStream();
+                byte[] arr = new byte[5000];
+                StreamUtil.readBytes(arr, length);
+                question = new String(arr, 0, length);
+                
+                if (question.startsWith("~:")) {
+                	question = question.substring(2);
+					idx = 0;
+
+					try {
+						idx = Integer.parseInt(question);
+					} catch (Exception var8) { }
+
+					question = recoveryQuestions[idx];
+				}
+                
+                Panel.setControlText(Client.panelRecovery, Client.controlRecoveryQuestion[var12], question);
+			}
+
+			if (Client.failedRecovery) {
+				Client.setLoginMessage("", "Sorry, you have already attempted 1 recovery, try again later");
+				return;
+			}
+
+			Client.login_screen = Client.SCREEN_PASSWORD_RECOVERY;
+			Panel.setControlText(Client.panelRecovery, Client.controlRecovery1,
+					"@yel@To prove this is your account please provide the answers to");
+			Panel.setControlText(Client.panelRecovery, Client.controlRecovery2,
+					"@yel@your security questions. You will then be able to reset your password");
+
+			for (int i = 0; i < 5; ++i) {
+				Panel.setControlText(Client.panelRecovery, Client.controlRecoveryInput[i], "");
+			}
+
+			Panel.setControlText(Client.panelRecovery, Client.recoverOldPassInput, "");
+			Panel.setControlText(Client.panelRecovery, Client.recoverNewPassInput, "");
+			Panel.setControlText(Client.panelRecovery, Client.recoverConfirmPassInput, "");
+			return;
+		} catch (Exception e) {
+			Client.setLoginMessage("Check your internet settings", "Sorry! Unable to connect to server.");
+			return;
+		}
+  }
+  
+  public static void recover() {
+	  Client.setLoginMessage("Connecting to server", "Please wait...");
+
+		try {
+			int port = Replay.connection_port;
+	        StreamUtil.initializeStream(Client.server_address, port);
+	        StreamUtil.setStreamMaxRetries(Client.maxRetries);
+	        
+	        Client.session_id = 0; // TODO: should read int to get session ID here, triggered by TCP handshake
+	        
+			String oldPass = Client.formatText(Panel.getControlText(Client.panelRecovery, Client.recoverOldPassInput), 20);
+			String newPass = Client.formatText(Panel.getControlText(Client.panelRecovery, Client.recoverNewPassInput), 20);
+			
+			StreamUtil.newPacket(8);
+	        Object buffer = StreamUtil.getStreamBuffer();
+	        StreamUtil.putLongTo(buffer, Util.username2hash(Panel.getControlText(Client.panelLogin, Client.loginUserInput)));
+			
+			// In 235 putRandom of "random.dat" is 24 bytes, but for 127 is expected 4 bytes.
+	        Object randBlock = StreamUtil.getNewBuffer(24);
+	        StreamUtil.putRandom(randBlock);
+	        byte[] randArr = StreamUtil.getBufferByteArray(randBlock);
+	        CRC16 sum = new CRC16();
+	        sum.update(randArr);
+	        int rand = (int) sum.getValue();
+	        StreamUtil.putIntTo(buffer, rand);

```

Comment:
I think this int is actually always 0? unless something external reaches in and modifies it...

Reply:
Probably they modified somewhere else or something, in 127 they send it as a seed, it being a public static could have easily been modified by launcher maybe

Reply:
Yeah, I think this implementation is probably correct. Resolving. Going to assume Launcher does inject uid.dat into link.uid. Even if not, it does it in 127, so it's authentic.